### PR TITLE
Fix some divisions by zero

### PR DESCRIPTION
**Pull request by illwieckz**
Before those patches, I was getting the following errors when doing `feenableexcept(FE_DIVBYZERO);`.

I added one patch that copy-pastes a fix from one code to another code that looks very close, as maybe it can happen too but would require some other testing scenes.

Clang 19, RelWithDebInfo build (yes, this one looks weird, but it disappears once the patches are applied, probably some optimization obfuscating the produced code too much for the debugger):

```
Thread 27 "daemon" received signal SIGFPE, Arithmetic exception.

0x00007fcdd6ef92b6 in G_RunFrame (levelTime=<optimized out>) at Unvanquished/src/sgame/sg_main.cpp:2252
2252		for ( i = 0; i < level.num_entities; i++, ent++ )

Thread 27 (Thread 0x7fcdd6c006c0 (LWP 1005469) "daemon"):
#0  0x00007fcdd6ef92b6 in G_RunFrame (levelTime=<optimized out>) at Unvanquished/src/sgame/sg_main.cpp:2252
```

```
Thread 28 "daemon" received signal SIGFPE, Arithmetic exception.

0x000077865e78eebf in CG_CalcColorGradingForPoint (loc=<optimized out>) at Unvanquished/src/cgame/cg_view.cpp:1418
1418		cg.refdef.gradingWeights[1] = haveGlobal ? ( 1.0f - totalWeight ) : ( selectedWeight[0] / totalWeight );

Thread 28 (Thread 0x77868cc006c0 (LWP 1383380) "daemon"):
#0  0x000077865e78eebf in CG_CalcColorGradingForPoint (loc=<optimized out>) at Unvanquished/src/cgame/cg_view.cpp:1418
#1  CG_CalcViewValues () at Unvanquished/src/cgame/cg_view.cpp:1685
#2  CG_DrawActiveFrame (serverTime=<optimized out>, demoPlayback=<optimized out>) at Unvanquished/src/cgame/cg_view.cpp:1966
```

Clang 19, Debug build:

```
Thread 28 "daemon" received signal SIGFPE, Arithmetic exception.

0x000074981277eb79 in CG_AddReverbEffects (loc=0x749812bbb2f4 <cg+1140>) at Unvanquished/src/cgame/cg_view.cpp:1554
1554			weight = 1.0f - dist / cgs.gameReverbDistances[i];

Thread 28 (Thread 0x7498374006c0 (LWP 1038016) "daemon"):
#0  0x000074981277eb79 in CG_AddReverbEffects (loc=0x749812bbb2f4 <cg+1140>) at Unvanquished/src/cgame/cg_view.cpp:1554
#1  0x000074981277d8e7 in CG_CalcViewValues () at Unvanquished/src/cgame/cg_view.cpp:1687
#2  0x000074981277d6bf in CG_DrawActiveFrame (serverTime=1125, demoPlayback=false) at Unvanquished/src/cgame/cg_view.cpp:1965
```

GCC 14, RelWithDebInfo build:

```
Thread 27 "daemon" received signal SIGFPE, Arithmetic exception.

0x000076ecef710776 in G_EvaluateAcceleration (ent=<optimized out>, msec=<optimized out>) at Unvanquished/src/sgame/sg_main.cpp:2151
2151		VectorScale( deltaVelocity, 1.0f / ( float ) msec, ent->acceleration );

Thread 27 (Thread 0x76ed07c006c0 (LWP 1337649) "daemon"):
#0  0x000076ecef710776 in G_EvaluateAcceleration (ent=<optimized out>, msec=<optimized out>) at Unvanquished/src/sgame/sg_main.cpp:2151
#1  G_RunFrame (levelTime=<optimized out>) at Unvanquished/src/sgame/sg_main.cpp:2283
#2  0x000076ecef6c5924 in operator() (__closure=<optimized out>, levelTime=<optimized out>) at Unvanquished/src/sgame/sg_api.cpp:131
```

```
Thread 42 "daemon" received signal SIGFPE, Arithmetic exception.

0x0000790e5276263b in CG_CalculateTimeFrac (birth=<optimized out>, life=<optimized out>, delay=<optimized out>) at Unvanquished/src/cgame/cg_particles.cpp:2126

Thread 42 (Thread 0x790ed8c006c0 (LWP 1831090) "daemon"):
#0  0x0000790e5276263b in CG_CalculateTimeFrac (birth=<optimized out>, life=<optimized out>, delay=<optimized out>) at Unvanquished/src/cgame/cg_particles.cpp:2126
#1  CG_RenderParticle (p=p@entry=0x790e53e55b40 <particles+1152>) at Unvanquished/src/cgame/cg_particles.cpp:2521
#2  0x0000790e527655f7 in CG_AddParticles () at Unvanquished/src/cgame/cg_particles.cpp:2668
#3  0x0000790e527ac215 in CG_DrawActiveFrame (serverTime=<optimized out>, demoPlayback=<optimized out>) at Unvanquished/src/cgame/cg_view.cpp:2001
#4  0x0000790e5272b201 in operator() (__closure=<optimized out>, serverTime=<optimized out>, demoPlayback=<optimized out>) at Unvanquished/src/cgame/cg_api.cpp:88
```

### CODE REVIEW

**Comment by illwieckz** [comment]
Also when using `feenableexcept (FE_OVERFLOW);` I get this when attempting to render the pla23 alien base, but I don't know how to fix that, here is Clang 19, RelWithDebInfo build. I reproduce the same with GCC 14.

```
Thread 27 "daemon" received signal SIGFPE, Arithmetic exception.

0x0000774382d22cf8 in VectorLengthSquared (v=<optimized out>) at Unvanquished/daemon/src/engine/qcommon/q_shared.h:570
570		return v[ 0 ] * v[ 0 ] + v[ 1 ] * v[ 1 ] + v[ 2 ] * v[ 2 ];

Thread 27 (Thread 0x774382a006c0 (LWP 2174030) "daemon"):
#0  0x0000774382d22cf8 in VectorLengthSquared (v=<optimized out>) at Unvanquished/daemon/src/engine/qcommon/q_shared.h:570
#1  DistanceSquared (p1=0x7743693a014c, p2=<optimized out>) at Unvanquished/daemon/src/engine/qcommon/q_shared.h:589
#2  GetCloseLocationEntity (ent=0x7743693a0000) at Unvanquished/src/sgame/sg_team.cpp:369
#3  CheckTeamStatus () at Unvanquished/src/sgame/sg_team.cpp:559
#4  0x0000774382cf96f5 in G_RunFrame (levelTime=<optimized out>) at Unvanquished/src/sgame/sg_main.cpp:2382
```

**Comment by illwieckz** [comment]
I also get this at the same place when using `feenableexcept (FE_UNDERFLOW);`, but this happens when loading the map:

```
Thread 1 "daemon" received signal SIGFPE, Arithmetic exception.

0x00005ebc13147ea5 in VectorLengthSquared (v=0x7ffd680d5658) at Unvanquished/daemon/src/engine/qcommon/q_shared.h:570
570		return v[ 0 ] * v[ 0 ] + v[ 1 ] * v[ 1 ] + v[ 2 ] * v[ 2 ];

Thread 1 (Thread 0x71850ea98a80 (LWP 2244806) "daemon"):
#0  0x00005ebc13147ea5 in VectorLengthSquared (v=0x7ffd680d5658) at Unvanquished/daemon/src/engine/qcommon/q_shared.h:570
#1  R_TBNtoQtangents (tangent=0x7184f1666d60, binormal=0x7184f1666d70, normal=0x7184f1666d80, qtangent=qtangent@entry=0x7184f166dd70) at Unvanquished/daemon/src/engine/renderer/tr_main.cpp:197
#2  0x00005ebc1315916b in R_GenerateMD5VBOSurface (surfName=..., vboTriangles=std::vector of length 452, capacity 1000 = {...}, md5=md5@entry=0x71850cc67aa0, surf=surf@entry=0x71850cc679a0, skinIndex=<optimized out>, boneReferences=0x7ffd680d5900) at Unvanquished/daemon/src/engine/renderer/tr_model_skel.cpp:158
#3  0x00005ebc131570c9 in R_LoadMD5 (mod=mod@entry=0x71850cc67b00, buffer=<optimized out>, modName=0x5ebc1636f3e0 "models/weapons/grenade/grenade.md5mesh") at Unvanquished/daemon/src/engine/renderer/tr_model_md5.cpp:632
#4  0x00005ebc131508b8 in RE_RegisterModel (name=0x5ebc1636f3e0 "models/weapons/grenade/grenade.md5mesh") at Unvanquished/daemon/src/engine/renderer/tr_model.cpp:152
```

**Comment by VReaperV** [comment]
> _Edit:_ In `GetCloseLocationEntity()` all components of `eloc->r.currentOrigin` are `1.70000001e+19`…

Is it the fake entity maybe?

**Comment by illwieckz** [comment]
Ah,, very likely:

```c++
void G_SpawnFakeEntities()
{
	level.fakeLocation = G_NewEntity( NO_CBSE );
	level.fakeLocation->s.origin[ 0 ] =
	level.fakeLocation->s.origin[ 1 ] =
	level.fakeLocation->s.origin[ 2 ] = 1.7e19f; // well out of range
```

**Comment by illwieckz** [comment]
I added a `BEST_LEN` constant and use a multiple of it instead. Though it is currently a define and if we prefer an `extern float` I don't know where to store it (in `bg_pmove.cpp` maybe?).

**Comment by illwieckz** [comment]
I also get this overflow in engine, when playing the [0.55.1 benchmark demo](https://dl.unvanquished.net/benchmark/demos/unvanquished-benchmark_0.55.1.dm_86.xz).

```
Thread 1 "daemon" received signal SIGFPE, Arithmetic exception.
0x00005d2a1e9960eb in DotProduct (x=0x7ffd0af18f38, y=0x7ffd0af18f38) at Unvanquished/daemon/src/engine/qcommon/q_shared.h:375
375		return x[ 0 ] * y[ 0 ] + x[ 1 ] * y[ 1 ] + x[ 2 ] * y[ 2 ];

Thread 1 (Thread 0x7dbeeda89a80 (LWP 3515290) "daemon"):
#0  0x00005d2a1e9960eb in DotProduct (x=0x7ffd0af18f38, y=0x7ffd0af18f38) at Unvanquished/daemon/src/engine/qcommon/q_shared.h:375
#1  VectorNormalize (v=0x7ffd0af18f38) at Unvanquished/daemon/src/engine/qcommon/q_shared.h:602
#2  RE_LerpTagET (tag=0x7ffd0af18f14, refent=0x7ffd0af1f070, tagNameIn=<optimized out>, startIndex=<optimized out>) at Unvanquished/daemon/src/engine/renderer/tr_model.cpp:501
```

File: src/shared/bg_public.h [hunk 1]
```diff
@@ -82,6 +81,8 @@ struct glm4rw_impl
 
 #include "engine/qcommon/q_shared.h"
 
+#define BEST_LEN 3.0f * 8192.0f * 8192.0f
```
Comment (slipher) [inline src/shared/bg_public.h:84]: The naming of this makes no sense. What is it supposed to mean?

You could make a constant named like HUGE_DISTANCE, make it 1e10 or 1e15. That's big enough that no one could confuse it with a normal distance, but small enough that it won't cause overflows. It would be good to replace the existing instances of HUGE_QFLT (which are all distances) with that, since HUGE_QFLT is too big and has the overflow risk.

File: src/cgame/cg_view.cpp [hunk 2]
```diff
@@ -1414,10 +1415,27 @@ static void CG_CalcColorGradingForPoint( vec3_t loc )
...
+		cg.refdef.gradingWeights[3] = selectedWeight[2];
+
+		cg.refdef.gradingWeights[0] /= totalWeight;
+		cg.refdef.gradingWeights[1] /= totalWeight;
+		cg.refdef.gradingWeights[2] /= totalWeight;
+		cg.refdef.gradingWeights[3] /= totalWeight;
+	}
+	else
+	{
+		Vector4Set( cg.refdef.gradingWeights, 0, 0, 0, 0 );
```
Comment (slipher) [inline src/cgame/cg_view.cpp:1432]: This case should not exist because it would cause the screen to be rendered completely black.

Comment (illwieckz) [reply src/cgame/cg_view.cpp:1432]: I don't know what that code does, but I have noticed the next test prevents it to be rendered black. So the original behavior (minus the division by zero) seem to have been conserved.

Comment (illwieckz) [reply src/cgame/cg_view.cpp:1432]: I just rewrote the code a bit because I didn't want to do some hard-to-read `a ? b : c ? d : e` and while at it I rewrote it in a way it helps the compiler to vectorize it, but I expect the code to do the same as before, minus the division by zero.

File: src/shared/bg_public.h (hunk 1 above)
Comment (illwieckz) [reply src/shared/bg_public.h:84]: I only named it `BEST_LEN` because it was named `bestlen` before, so “not my fault!”. 🤪️

I welcome any better name.

Comment (illwieckz) [reply src/shared/bg_public.h:84]: Hmm, `HUGE_QFLT` is defined in engine but only used in sgame, so we may move it to the game.

I wonder what can mean `QFLT`.

Also I don't know what this comment means about `MAX_QFLT` (that doesn't exist):

```c++
#define HUGE_QFLT 3e38f // TODO: Replace HUGE_QFLT with MAX_QFLT
```

Comment (illwieckz) [reply src/shared/bg_public.h:84]: Maybe QFLT means some QFloat like some QBoolean.

Comment (illwieckz) [reply src/shared/bg_public.h]: Maybe the comment about `MAX_QFLT` was not about using the max float value for `MAX_QFLT`, but about giving us a new `max` to use that is not the actual max of the float (so, one that doesn't overflow).

File: src/cgame/cg_view.cpp (hunk 2 above)
Comment (slipher) [reply src/cgame/cg_view.cpp:1432]: I managed to find a map where it can be tested, so I rewrote that code. https://github.com/Unvanquished/Unvanquished/pull/3213

**Comment by illwieckz** [comment]
I just caught a very weird division by zero error  when building with Clang 19. All I have to do is to comment out `set_c_cxx_flag("-ffast-math")` in `DaemonFlags.cmake` and add some `#include <cfenv>` and  `feenableexcept(FE_DIVBYZERO);` code.

Then I catch that:

```
Thread 29 "daemon" received signal SIGFPE, Arithmetic exception.

0x000073564a721ab3 in CG_LoadBeaconsConfig () at Unvanquished/src/cgame/cg_beacon.cpp:157
157		bc->fadeMinDist = Square( bc->hudSize / bc->hudMaxSize );

Thread 29 (Thread 0x735678c006c0 (LWP 1181145) "daemon"):
#0  0x000073564a721ab3 in CG_LoadBeaconsConfig () at Unvanquished/src/cgame/cg_beacon.cpp:157
#1  0x000073564a73e1f8 in CG_Init (serverMessageNum=2, clientNum=0, gl=..., gameState=...) at Unvanquished/src/cgame/cg_main.cpp:1363
```

The problem is that if I modify that function to add some loggers this way:

```c++
	Log::Warn("size: %d", bc->hudSize);
	Log::Warn("max: %d", bc->hudMaxSize);
	Log::Warn("min: %d", bc->hudMinSize);

	bc->fadeMinDist = Square( bc->hudSize / bc->hudMaxSize );
	bc->fadeMaxDist = Square( bc->hudSize / bc->hudMinSize );
```

Here is what is printed:

```
Warn: size: 1440 
Warn: max: 1008 
Warn: min: 36 
```

But `Square` is defined this way in `q_shared.h`:

```c++
#define Square( x ) ( ( x ) * ( x ) )
```

So basically the division by zero happens when doing that:

```c++
bc->fadeMinDist = ( ( 1440 / 1008 ) * ( 1440 / 1008 ) );
```

That doesn't make sense, there is no division by zero here.

I don't get this error with `Debug` build, only with `DebWithRelInfo` one.

**Comment by illwieckz** [comment]
If I do that, I don't get the division by zero:

```c++
	Log::Warn("size: %d", bc->hudSize);
	Log::Warn("max: %d", bc->hudMaxSize);
	Log::Warn("min: %d", bc->hudMinSize);
	
	int a = bc->hudSize / bc->hudMaxSize;
	int b = Square( a );

	bc->fadeMinDist = b;
	bc->fadeMaxDist = Square( bc->hudSize / bc->hudMinSize );
```

But then I got one later:

```
Thread 19 "render thread" received signal SIGFPE, Arithmetic exception.

0x00005ac251e00b0f in RB_RenderSSAO () at Unvanquished/daemon/src/engine/renderer/tr_backend.cpp:3230
3230		zParams[ 0 ] = 2.0f * tanf( DEG2RAD( backEnd.refdef.fov_x * 0.5f ) ) / glConfig.vidWidth;

#0  0x00005ac251e00b0f in RB_RenderSSAO () at Unvanquished/daemon/src/engine/renderer/tr_backend.cpp:3230
#1  0x00005ac251e04045 in RB_RenderView (depthPass=<optimized out>) at Unvanquished/daemon/src/engine/renderer/tr_backend.cpp:4703
#2  0x00005ac251e03d5e in DrawViewCommand::ExecuteSelf (this=0x701c8151aff4) at Unvanquished/daemon/src/engine/renderer/tr_backend.cpp:5567
#3  0x00005ac251e07589 in RB_ExecuteRenderCommands (data=0x701c81519d2c) at Unvanquished/daemon/src/engine/renderer/tr_backend.cpp:5813
#4  RB_RenderThread () at Unvanquished/daemon/src/engine/renderer/tr_backend.cpp:5844
#5  0x00005ac251ea65d2 in GLimp_RenderThreadWrapper () at Unvanquished/daemon/src/engine/sys/sdl_glimp.cpp:224
#6  0x0000701c85eaa296 in SDL_RunThread (thread=<optimized out>) at /usr/src/libsdl2-2.30.0+dfsg-1build3/src/thread/SDL_thread.c:292
#7  RunThread (data=0x5ac254fba8b0) at /usr/src/libsdl2-2.30.0+dfsg-1build3/src/thread/pthread/SDL_systhread.c:76
```

But then if I do that:

```c++
	Log::Warn("width: %d", glConfig.vidWidth);
	Log::Warn("width: %d", glConfig.vidHeight);
	zParams[ 0 ] = 2.0f * tanf( DEG2RAD( backEnd.refdef.fov_x * 0.5f ) ) / glConfig.vidWidth;
	zParams[ 1 ] = 2.0f * tanf( DEG2RAD( backEnd.refdef.fov_y * 0.5f ) ) / glConfig.vidHeight;
```

It prints this before crashing:

```
Warn: width: 2560 
Warn: width: 1440
```

There is no division by zero at all!

**Comment by illwieckz** [comment]
Hmm, that should be floats, so if I do that:

```c++
	float a = bc->hudSize / bc->hudMaxSize;
	float b = Square( a );
	
	bc->fadeMinDist = b;
```

I also get the division by zero error:

```
Thread 29 "daemon" received signal SIGFPE, Arithmetic exception.

0x0000768f1e721b2b in CG_LoadBeaconsConfig () at Unvanquished/src/cgame/cg_beacon.cpp:161

161		float a = bc->hudSize / bc->hudMaxSize;
```

That doesn't make sense at all…

**Comment by illwieckz** [comment]
My tree reproducing the bug:

```diff
diff --git a/src/engine/framework/System.cpp b/src/engine/framework/System.cpp
index 593b94e20..a3c30be39 100644
--- a/src/engine/framework/System.cpp
+++ b/src/engine/framework/System.cpp
@@ -28,6 +28,10 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ===========================================================================
 */
 
+#ifdef __linux__
+#include <fenv.h>
+#endif
+
 #include "qcommon/q_shared.h"
 #include "qcommon/qcommon.h"
 #include "Application.h"
@@ -740,6 +744,10 @@ __declspec(dllexport) void DummyPreventingLinkerFromBreakingASLR() {}
 // This is why ALIGN_STACK_FOR_MINGW is needed (normally gcc would generate alignment code in main()).
 ALIGN_STACK_FOR_MINGW int main(int argc, char** argv)
 {
+#ifdef __linux__
+       feenableexcept(FE_DIVBYZERO);
+#endif
+
        // Initialize the engine. Any errors here are fatal.
        try {
                Sys::Init(argc, argv);

diff --git a/cmake/DaemonFlags.cmake b/cmake/DaemonFlags.cmake
index fbf3637a8..73bb45014 100644
--- a/cmake/DaemonFlags.cmake
+++ b/cmake/DaemonFlags.cmake
@@ -238,7 +238,7 @@ else()
        endif()
 
        # Optimizations.
-       set_c_cxx_flag("-ffast-math")
+#      set_c_cxx_flag("-ffast-math")
 
        # Use hidden symbol visibility if possible.
        try_c_cxx_flag(FVISIBILITY_HIDDEN "-fvisibility=hidden")
```

DLL `RelWithDebInfo` build.

**Comment by illwieckz** [comment]
I reproduce with Clang 19, 18, 17, 16, 15, but not Clang 14.

**Comment by illwieckz** [comment]
Full reproduction instructions:

```sh
cat > /tmp/fenv.patch <<\EOF
diff --git a/cmake/DaemonFlags.cmake b/cmake/DaemonFlags.cmake
index fbf3637a8..73bb45014 100644
--- a/cmake/DaemonFlags.cmake
+++ b/cmake/DaemonFlags.cmake
@@ -238,7 +238,7 @@ else()
 	endif()
 
 	# Optimizations.
-	set_c_cxx_flag("-ffast-math")
+#	set_c_cxx_flag("-ffast-math")
 
 	# Use hidden symbol visibility if possible.
 	try_c_cxx_flag(FVISIBILITY_HIDDEN "-fvisibility=hidden")
diff --git a/src/engine/framework/System.cpp b/src/engine/framework/System.cpp
index 593b94e20..a64a819b2 100644
--- a/src/engine/framework/System.cpp
+++ b/src/engine/framework/System.cpp
@@ -28,6 +28,7 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ===========================================================================
 */
 
+#include <cfenv>
 #include "qcommon/q_shared.h"
 #include "qcommon/qcommon.h"
 #include "Application.h"
@@ -740,6 +741,7 @@ __declspec(dllexport) void DummyPreventingLinkerFromBreakingASLR() {}
 // This is why ALIGN_STACK_FOR_MINGW is needed (normally gcc would generate alignment code in main()).
 ALIGN_STACK_FOR_MINGW int main(int argc, char** argv)
 {
+	feenableexcept(FE_DIVBYZERO);
 	// Initialize the engine. Any errors here are fatal.
 	try {
 		Sys::Init(argc, argv);
EOF

git fetch origin
git checkout 6df472fa84a0314bbc1cf0402fd9dea295ac9b45
git submodule update --init --recursive
git checkout illwieckz/divby0
(cd daemon ; patch -p1 < /tmp/fenv.patch)
cmake -Bbuild/fenv . -DCMAKE_C_COMPILER=clang-19 -DCMAKE_CXX_COMPILER=clang++-19 -DCMAKE_BUILD_TYPE=RelWithDebInfo -DUSE_EXTERNAL_DEPS=OFF -DBUILD_SERVER=OFF -DBUILD_TTY_CLIENT=OFF
cmake --build build/fenv --parallel "$(nproc)"
gdb -ex run -args build/fenv/daemon -set vm.cgame.type 3 -set vm.sgame.type 3 -pakpath "${XDG_DATA_HOME:-${HOME}/.local/share}/unvanquished/base/pkg" +devmap plat23
```

**Comment by illwieckz** [comment]
With LLDB:

```
* thread #23, name = 'daemon', stop reason = signal SIGFPE: floating point divide by zero
    frame #0: 0x00007fff6a852c03 cgame-native-dll.so`CG_LoadBeaconsConfig() at cg_beacon.cpp:157:20
   154 				Log::Warn( "bad keyword \"%s\" in \"%s\"", token.string, path );
   155 		}
   156 	
-> 157 		bc->fadeMinDist = Square( bc->hudSize / bc->hudMaxSize );
   158 		bc->fadeMaxDist = Square( bc->hudSize / bc->hudMinSize );
   159 	
   160 		Parse_FreeSourceHandle( fd  );
(lldb) bt 23
* thread #23, name = 'daemon', stop reason = signal SIGFPE: floating point divide by zero
  * frame #0: 0x00007fff6a852c03 cgame-native-dll.so`CG_LoadBeaconsConfig() at cg_beacon.cpp:157:20
    frame #1: 0x00007fff6a86cf18 cgame-native-dll.so`CG_Init(serverMessageNum=2, clientNum=0, gl=0x00007fff98bfedd0, gameState=0x00007fff98bf6dd0) at cg_main.cpp:1363:2
    frame #2: 0x00007fff6a84feff cgame-native-dll.so`VM::VMHandleSyscall(unsigned int, Util::Reader) [inlined] VM::VMHandleSyscall(unsigned int, Util::Reader)::$_1::operator()(this=<unavailable>, serverMessageNum=<unavailable>, clientNum=<unavailable>, gl=0x00007fff98bfedd0, gamestate=0x00007fff98bf6dd0) const at cg_api.cpp:62:21
```

**Comment by illwieckz** [comment]
The disassembled code of such division is (snippet):

```asm
   0x00007fff72852bee <+1598>:  movups 0x39998(%r15),%xmm0
   0x00007fff72852bf6 <+1606>:  movsd  0x3999c(%r15),%xmm1
   0x00007fff72852bff <+1615>:  shufps $0x0,%xmm0,%xmm0
=> 0x00007fff72852c03 <+1619>:  divps  %xmm1,%xmm0
```

I wonder if the division by zero comes from the unused fields of the xmm registers… (that would be very stupid 🤦‍♀️️)

**Comment by illwieckz** [comment]
Doing this:

```c++
bc->fadeMinDist = Square( (double long) bc->hudSize / (double long) bc->hudMaxSize );
```

and then avoiding the usage of the xmm register (I guess), works around the bug.

**Comment by illwieckz** [comment]
This also works around the bug:

```
diff --git a/src/cgame/cg_local.h b/src/cgame/cg_local.h
index b281f33e8..fa68930c3 100644
--- a/src/cgame/cg_local.h
+++ b/src/cgame/cg_local.h
@@ -711,8 +711,17 @@ struct beaconsConfig_t
 
 	// HUD
 	float         hudSize;
+	float a01 = 1.0f;
+	float a02 = 1.0f;
+	float a03 = 1.0f;
 	float         hudMinSize;
+	float a11 = 1.0f;
+	float a12 = 1.0f;
+	float a13 = 1.0f;
 	float         hudMaxSize;
+	float a21 = 1.0f;
+	float a22 = 1.0f;
+	float a23 = 1.0f;
 	float         hudAlpha;
 	float         hudAlphaImportant;
 	vec2_t        hudCenter;    //runtime
```

So I'm now seriously considering we get a division by zero error from the unused fields of the xmm registers.

**Comment by illwieckz** [comment]
Small file to reproduce the bug:

Save this as `broken-xmm-div0.cpp`:

```c++
#include <cfenv>
#include <cstdio>

#define Square( x ) ( ( x ) * ( x ) )

struct beaconsConfig_t
{
	float fadeMinDist;
	float fadeMaxDist;

	float hudSize;
	float a01 = 0.0f;
	float a02 = 0.0f;
	float a03 = 0.0f;
	float hudMinSize;
	float a11 = 0.0f;
	float a12 = 0.0f;
	float a13 = 0.0f;
	float hudMaxSize;
	float a21 = 0.0f;
	float a22 = 0.0f;
	float a23 = 0.0f;
};

void CG_LoadBeaconsConfig( beaconsConfig_t *bc );

void CG_LoadBeaconsConfig( beaconsConfig_t *bc )
{
	feenableexcept(FE_DIVBYZERO);
	bc->fadeMinDist = Square( bc->hudSize / bc->hudMaxSize );
	bc->fadeMaxDist = Square( bc->hudSize / bc->hudMinSize );
}

int main( int , char ** )
{
	beaconsConfig_t bc;

	sscanf( "1440 1008", "%f %f", &bc.hudSize, &bc.hudMaxSize);
	CG_LoadBeaconsConfig( &bc );
	printf("%f\n", bc.fadeMinDist);

	return 0;
}

```

Then run:

```sh
clang++ -std=gnu++14 -march=x86-64 -O2 -g3 broken-xmm-div0.cpp -o broken-xmm-div0
lldb -o run -o quit ./broken-xmm-div0
```

**Comment by illwieckz** [comment]
I reported the false positive `divisions by zero` exceptions on clang here:

- https://github.com/llvm/llvm-project/issues/118265

**Comment by illwieckz** [comment]
So, I took the opportunity to do extra research on some underflow or to investigate the model bug, but this PR looks ready to me.

File: src/cgame/cg_view.cpp [hunk 3]
```diff
@@ -1414,10 +1414,27 @@ static void CG_CalcColorGradingForPoint( vec3_t loc )
...
 	}
 
-	cg.refdef.gradingWeights[0] = 0.0f;
-	cg.refdef.gradingWeights[1] = haveGlobal ? ( 1.0f - totalWeight ) : ( selectedWeight[0] / totalWeight );
-	cg.refdef.gradingWeights[2] = totalWeight == 0.0f ? 0.0f : selectedWeight[1] / totalWeight;
-	cg.refdef.gradingWeights[3] = totalWeight == 0.0f ? 0.0f : selectedWeight[2] / totalWeight;
+	if ( totalWeight )
+	{
+		cg.refdef.gradingWeights[0] = 0.0f;
+		cg.refdef.gradingWeights[1] = selectedWeight[0];
```
Comment (VReaperV) [inline src/cgame/cg_view.cpp:1421]: Why not do the divisions here?

File: src/cgame/cg_view.cpp [hunk 4]
```diff
@@ -1414,10 +1414,27 @@ static void CG_CalcColorGradingForPoint( vec3_t loc )
...
+		cg.refdef.gradingWeights[3] /= totalWeight;
+	}
+	else
+	{
+		Vector4Set( cg.refdef.gradingWeights, 0, 0, 0, 0 );
+	}
+
+	if ( haveGlobal )
+	{
+		cg.refdef.gradingWeights[1] = 1.0f - totalWeight ;
```
Comment (VReaperV) [inline src/cgame/cg_view.cpp:1437]: Why is this needed now?

Comment (VReaperV) [inline src/cgame/cg_view.cpp:1437]: Extraneous whitespace.

File: src/cgame/cg_view.cpp (hunk 3 above)
Comment (illwieckz) [reply src/cgame/cg_view.cpp]: This code will be obsoleted by: https://github.com/Unvanquished/Unvanquished/pull/3213

File: src/cgame/cg_view.cpp (hunk 4 above)
Comment (illwieckz) [reply src/cgame/cg_view.cpp]: This code will be obsoleted by: https://github.com/Unvanquished/Unvanquished/pull/3213

The code was doing that before, but without a division by zero for free. I don't know what this code does.

Comment (illwieckz) [reply src/cgame/cg_view.cpp]: This code will be obsoleted by: https://github.com/Unvanquished/Unvanquished/pull/3213

File: src/cgame/cg_view.cpp [hunk 5]
```diff
@@ -1551,8 +1551,8 @@ static void CG_AddReverbEffects( vec3_t loc )
 		}
 
 		dist = CM_DistanceToModel( loc, cgs.gameReverbModels[i] );
-		weight = 1.0f - dist / cgs.gameReverbDistances[i];
-		weight = Math::Clamp( weight, 0.0f, 1.0f ); // Maths::clampFraction( weight )
+
+		weight = 1.0f - dist / std::max( 0.01f, cgs.gameReverbDistances[i] );
```
Comment (illwieckz) [inline src/cgame/cg_view.cpp:1555]: I rewrote it in the same ways it has been done for `cgs.gameGradingDistances` in:

- https://github.com/Unvanquished/Unvanquished/pull/3213

File: src/cgame/cg_particles.cpp [hunk 6]
```diff
@@ -2123,9 +2123,8 @@ Calculate the fraction of time passed
 */
 static float CG_CalculateTimeFrac( int birth, int life, int delay )
 {
-	float frac = ( ( float ) cg.time - ( float )( birth + delay ) ) / ( float )( life - delay );
-
-	return Math::Clamp( frac, 0.0f, 1.0f );
+	int diff = life - delay;
+	return diff ? Math::Clamp( (float) (cg.time - birth + delay) / (float) diff, 0.0f, 1.0f ) : 1.0f;
```
Comment (illwieckz) [inline src/cgame/cg_particles.cpp:2127]: I wonder if the clamp was a workaround for `infinity` and if it is required.

**Review by illwieckz — COMMENTED** [review]
I removed the colorgrading code in favor of:

- https://github.com/Unvanquished/Unvanquished/pull/3213

File: src/cgame/cg_particles.cpp (hunk 6 above)
Comment (VReaperV) [reply src/cgame/cg_particles.cpp:2127]: It looks like `( cg.time - birth + delay ) / (float) diff` can be > 1.0. E. g: `life = 10; delay = 3`, start at `cg.time = 0`, then at `cg.time = 8`: `diff = 7; cg.time - birth + delay = 11`, then the fraction is ~1.57.

Comment (VReaperV) [reply src/cgame/cg_particles.cpp:2127]: Technically, I think we also don't check that `life - delay > 0`, so it could be negative, but that shouldn't be a valid particle system. Doing that check would also skipping the `diff` check here.

Comment (illwieckz) [reply src/cgame/cg_particles.cpp:2127]: We may do that instead?

```c++
return diff > 0 ? Math::Clamp( (float) (cg.time - birth + delay) / (float) diff, 0.0f, 1.0f ) : 1.0f;
```

Comment (VReaperV) [reply src/cgame/cg_particles.cpp:2127]: I suppose. I don't think it matters all that much because, again, a particle system where `diff` is < 0 shouldn't be valid anyway and should never spawn.
